### ISSUE TITLE
l3seq: provide a method to set a term at a specific index

### ISSUE THREAD

**Issue by loopspace** [latex3/latex3#844]
Background: in the [hobby](https://github.com/loopspace/hobby) package there are a fair few calculations on vectors (and matrices).  In order to do the calculations, I need an array-like data structure to hold the vectors.  Specifically, I need to be able to get and set items at a specific index in an ordered list.

When I originally wrote that package, `\seq_item:Nn` didn't exist.  So I made a wrapper around a `prop` which, in essence, added the ordering part.  Now that `\seq_item:Nn` exists, a `seq` is _almost_ the data structure that I need.  What is missing is the ability to _set_ a specific term of the sequence.

Hence my request for a method to set an item at a specific index of a `seq`.

References:
* Issue on hobby repository: https://github.com/loopspace/hobby/issues/8
* A possible implementation on TeX-SX: https://tex.stackexchange.com/questions/497912/how-to-create-array-of-dims

**Comment by blefloch** [comment]
I'm not super active these days, so just a quick comment: have you looked at the `l3intarray` and `l3fparray` modules?  They allow manipulations of arrays of integers (secretely dimensions, so limited to (slightly less than) 31 bits), and decimal floating points, respectively.  I'm pretty sure that `l3fparray` could be improved, however.

**Comment by loopspace** [comment]
I would need to use `l3fparray` of those two.

The main issue with that is that I would need to keep track of a "last element" since I calculate the vectors incrementally, so I'd need to effectively implement an `fparray_push` function.  I guess I could allocate the first entry as the pointer to the current last one so that I didn't have extra variables, but I'd still need to define some extra routines.

The other thing missing from `l3fparray` is the ability to clear an array.  So either I'd have to have a new array for each time the algorithm was used, or I'd have to pick a huge size initially and re-use it, and hope that I've picked a size bigger than anyone is likely to need (but if I pick one that big, am I needlessly using up memory)?

So in conclusion, `fparray` certainly is promising but would still need a bit of extra stuff around it to make it useable for me.

**Comment by PhelypeOleinik** [comment]
Okay, some empirical tests I made show that in pdfTeX, loading an intarray takes about 300+`<num items>` words of font memory, out of a maximum of 8000000 words of memory for up to 9000 fonts (which can be increased of course, but let's forget that).  That means it's cheaper to declare one huge intarray than several small ones.  With those settings you have a bit more than 100 (supposing no one else uses them :) intarrays with 65536 items (taking Bruno's example) in pdfTeX, so some care should be taken.  And an fparray is three intarrays combined, so that number is shorter.

If your code uses one intarray, depending on how large you expect the biggest one to be, I think memory won't be a concern (with current LaTeX I needed a fparray with more than 2.5 million items to run out of memory).  How large would be the largest fparray you expect to be needed in your code? (also you can have an option to increase the allocation size, if a user needs more than the default).  I'm thinking this might be a good solution, but it would require to extend fparray (and intarray) to have push and pop functions (which might be more generally useful).  However a `\intarray_length:N` would be a bit strange, since the array is allocated with a fixed size, so maybe that you would need to keep track yourself.

---

Regardless of what @loopspace ends up using, I have needed some times to take specific items out of a `seq` variable or to put one at a specific position, so maybe something like `\seq_put_index:Nnn` and `\seq_get_index:NnN` would be good additions.

**Comment by car222222** [comment]
Going away from matrices etc!  So this maybe needs a distinct issue.

I also had on my list of useful extensions this type of enhancement to seq: something like \seq_put_index:Nnn and \seq_get_index:NnN.

Gong further, maybe implement on top of seq something along the lines of a 'general array' with entries of mixed types.  Like a 'pdf array' or lua table'.

**Comment by loopspace** [comment]
I guess part of my question is really "Is this something that might happen any time soon?".  With regard to the hobby package, redesigning it to use a better array structure than a `prop` will be a fairly major undertaking and so I'd rather only do it once.

I'm not trying to hurry the process here.  I have a couple of bugfixes for the hobby package that I've been sitting on for a while, and an interface to the `l3draw` package, so if this isn't something that will happen soon I may as well get those to CTAN now and put this on hold.  It'd just be helpful to know.  Thanks!

**Comment by blefloch** [comment]
> I guess part of my question is really "Is this something that might happen any
> time soon?".

I guess our silence on this indicates that it will not happen within a few
weeks, sorry.  At least on my side I'm completely underwater for 5 more weeks at
least.

**Comment by car222222** [comment]
But see comments on the commit.

**Comment by loopspace** [comment]
@car222222 Which commit are the relevant comments on?  I've just scanned through the ones that Bruno has linked above and can't see anything obvious.  As I'd like to use this in hobby, could you add a link to what you're referring to, please?

**Comment by loopspace** [comment]
Nevermind, I've found the comments now.

**Comment by car222222** [comment]
Sorry not on a commit!  
It is here: https://github.com/latex3/latex3/pull/891

**Comment by loopspace** [comment]
Thanks, so this issue is primarily as to the names of these things, meaning that I can start using them in the hobby code and then do a search-and-replace if the names change.

**Comment by car222222** [comment]
Maybe the idea of 'pop' for an array or seq item goes further than a naming issue.

What do you think this should do?  Remove the item completely, thus changing the location (index) of other items?  Or maybe change the value/entry of 'NoValue'?  (This second does not need 
an extra named command.)

Both behaviours seem strange to me.  

'popping' is fine for stacks, queues etc. (things that have distinguished locations) but not for general positions in general list structures.

**Comment by car222222** [comment]
Why not simply stick with Phelype's suggested names? :

\seq_put_index:Nnn and \seq_get_index:NnN would be good additions

I do not think that general seqs or arrays should allow any special values for items: hence no 'NoValue', please!

**Comment by u-fischer** [comment]
> What do you think this should do? Remove the item completely, thus changing the location (index) of other items? 

That is what `\seq_pop_item:NnN ` does. 

> Both behaviours seem strange to me.
>'popping' is fine for stacks, queues etc. 

We use `pop` also with properties for the "get + delete" action, so imho `pop` is okay as name for this operation too.

` \seq_get_index:NnN`  would be wrong as name, as get actions don't change the datatype.

**Comment by car222222** [comment]
"That is what \seq_pop_item:NnN does."

And that is something should never be done!  Would you do this to an array also?

**Comment by car222222** [comment]
"We use pop also with properties for the "get + delete" action"

We do this where?  You cannot sensibly "delete" any old item in a sequence or array.

Yes, "pop" normally means "get + delete" so should not be used on seq or array, etc. since there delete is not in general allowed.

**Comment by blefloch** [comment]
@LoopSpace: indeed, the functions won't disappear (well, they are in
l3candidates so if no one uses them they might, but I guess if you use them they
won't).  On the other hand, I think I can get a flexible version of intarray
implemented by the end of May, which might be better for you perhaps.


On 4/29/21 9:02 AM, Chris Rowley wrote:
> "That is what \seq_pop_item:NnN does."
> 
> And that is something should never be done! Would you do this to an array also?

Seq is much closer to a linked list than to an array.  It is true that you
wouldn't do this to an array.

On the other hand, Chris may be right about \seq_put_item being a bad name.  For
all other _put functions (\prop_put:Nnn, \seq_put_right:Nn) we can think of them
as typically adding an item, with the special case that \prop_put:Nnn removes
the old item if the key was already there.  In contrast \seq_put_item *never*
adds an item.  Perhaps name it \seq_set_item:Nnn?

**Comment by car222222** [comment]
Note that seq_pop_right is (just about:-) OK as it does not change the index of the remaining items.  

But pushing and popping at the left are to be frowned on or even prohibited (depending on what a sequence really is, of course).

**Comment by u-fischer** [comment]
> And that is something should never be done! Would you do this to an array also?

Why not?  A fast search  directly found lots of questions how to remove an entry from an list, and e.g. list.pop (in python) which does exactly this.  And we have other functions which shorten an seq, e.g. `\seq_remove_duplicates:N`.

**Comment by car222222** [comment]
Given the old pop possibilities for seq (left and right) the datatype would be better described as an 'indexed deque' (if such things exist:-).

**Comment by u-fischer** [comment]
> Perhaps name it \seq_set_item:Nnn?

I don't think that `set` would fit to the other uses of `set` in various places. I found `\seq_put_item` quite good, but `seq_replace_item` would perhaps work too.

**Comment by car222222** [comment]
If we can also remove and add items at any point then probably they should simply be called lists.  They then become essentially unindexed -- but if you keep track of what you do to,them then you could access a particular item by its position (or rather its 'last known position').  

These will be useless for many applications: so all this definitely means that for many common, and saner, uses a general array datatype is badly needed, something where the indexes of existing items are fixed for all time and things do not jump around (and cannot be popped into oblivion).

**Comment by car222222** [comment]
Yes, agreed: there is no problem for lists as in list.pop since they are not indexed.

**Comment by car222222** [comment]
@blefloch "Seq is much closer to a linked list than to an array."

, which has these two consequences:

1.  Do not pretend that seq is an indexed datatype.

2.  Provide a general 'indexed list' datatype with heterogeneous item values: maybe call it 'array':-) ?

**Comment by car222222** [comment]
\seq_remove_duplicates:N is probably a bad name (too late to change) since it is only the values that may be duplicated, the items never are since they all have distinct indices.

It is anyway a strange thing to do to any ordered datatype (any list).  

History is usually not a good guide to best practice.

**Comment by FrankMittelbach** [comment]
> > Perhaps name it \seq_set_item:Nnn?
> 
> I don't think that `set` would fit to the other uses of `set` in various places. I found `\seq_put_item` quite good, but `seq_replace_item` would perhaps work too.

Given that the command only allows you to change an existing item value (e.g., not allowing you to add new ones at the end of the sequence, say) I think `set_item` slightly preferable to `put` (because in other place we use put for both replace and add new and set always ofr setting  value for the bin has to exist already). `replace_item` would do too, but why  new name when "set" is used for this already elsewhere: I mean `\int_set:Nn` is really also "replace the value in register".

**Comment by FrankMittelbach** [comment]
> @blefloch "Seq is much closer to a linked list than to an array."
> 
> , which has these two consequences:
> 
>     1. Do not pretend that seq is an indexed datatype.
> 
>     2. Provide a general 'indexed list' datatype with heterogeneous item values: maybe call it 'array':-) ?

nobody is pretending that sequences are an indexed datatype with a permanent index (like an array) but that doesn't mean that you can't have a temporary index to provide direct access (including dropping an item at a specific location). Such an operation is not the norm or the major use but it is the very operation that was asked for (on seqs) and that is what got implemented.

**Comment by u-fischer** [comment]
@FrankMittelbach 
> replace_item would do too, but why new name 

It is not new, we have some `replace` already

> when "set" is used for this already elsewhere: 

I just wanted to claim,  `set` is only used for functions that affect a datatype as a whole. E.g. `\seq_set_from_clist:NN`, or `\prop_set_from_keyval:Nn` or `\clist_set_eq`, but we have also other uses. So for me it is fine.

**Comment by car222222** [comment]
Original request was for this:

 I need an array-like data structure to hold the vectors. 

and this:

Specifically, I need to be able to get and set items at a 
specific index in an ordered list.

And so it was not "including dropping an item at a specific location".
Thus pop (or any such deletion) is not apparently "the very operation that was asked for (on seqs)".

No popping was requested as the use case is for conventional arrays.

**Comment by car222222** [comment]
I could only find `_replace_` used for items in tl and str. 
They are there used for substrings/lists, which of course includes single items.

**Comment by car222222** [comment]
@blefloch If seq is to be a datatype for which deleting (popping) arbitrary items (referenced by index) is sensible, then I guess that they should also support 'pushing a new item (before and/or after) any indexed position'.

**Comment by blefloch** [comment]
There seems to be a few separate subjects:

- What is a seq?  Answer: a linked list; we don't call it that way because "list" is also used in document structure.  In many languages one can ask to remove/add items at a specific position of a linked list, even though this is not the main use case.  I agree we should also have a function such as `\seq_insert_item:Nnn` (sequence) (position) (item) which lengthens the sequence.

- Is `seq_pop_item` useful?  This is certainly debatable, I'm not sure.  Should I remove it?

- Name of setting one entry to some value?  Ulrike suggested `replace_item` but so far `replace` acts on values in the sense that it changes one value into another, rather than taking into account the position.  The objection that `set` normally changes the whole datatype is valid, but we already have `\intarray_set:Nnn` and `\char_set_catcode:Nn` which only affect one aspect of the object (intarray/char/…), so I think it is ok, and I now committed the name `set_item` accordingly.

- Can we have a fixed-size-array datatype?  Answer: not containing arbitrary TeX tokens (except by using one `\toks` per item, but there are too few `\toks` to make this viable), but I'm working on an extension of `intarray` that would allow arbitrary numeric types (int, dim, skip, muskip, fp).  In that "array" context, your objections to pop would definitely apply.  Future discussions on that could be moved to #882.

@car222222 It gets tough to follow when there are many separate comments, sorry if I missed something.

**Comment by car222222** [comment]
@blefloch There seems to be a few separate subjects: . . . 
It gets tough to follow when there are many separate comments, sorry if I missed something.

I do not think you did.  I am now suggesting moving these two:

1.  whether to add these to seq:
      remove,   + pop
      insert / push

into an issue on 

'Extension suggestions for seq, considered as a generic 
(linked) list'.

2.  Possible (feasible/useful) implementations of ‘fixed’ arrays (i.e. indexed lists).  Not sure if this should be an issue?  
For now, I shall make some remarks in an email.
